node.go

```diff
@@ -382,3 +390,17 @@ func (n *ProtoNode) Tree(p string, depth int) []string {
 	}
 	return out
 }
+
+func ProtoNodeConverter(b blocks.Block, nd ipld.Node) (legacy.UniversalNode, error) {
```

Comment:
The main use of this code is actually here: https://github.com/ipfs/go-merkledag/pull/67/files#diff-5e7919ad01b60e6e687fb533052ee768a19c08aa341016c5d2066f539c89bf52R28

Which feeds into here: https://github.com/ipfs/go-ipld-legacy/blob/master/decode.go#L44

The basic problem again is we need a node and block to satisfy the old interface. In the case of interacting with the block service in the way we do in go-merkledag, we end up with a block first. We then assemble a prime node from the block, the wrap both in a legacy node structure of some kind. The reason for the registry is that the ProtoNode (a legacy node implementation) in this repository has to have some special abilities -- namely that you can add and remove links and reserialize it. 

BTW, upon further reflection I believe I can remove the usage of this code from the go-path PR in go-ipfs.

The reason for having a cached copy of the the marshalled data in DagPB is that this struct is mutable, so you might mutate and remarshal. In ipld-prime, a node is never mutable -- it's a seperate interface from a nodebuilder, which is how you can actually construct and mutate nodes. I don't see why you would ever keep marshalled data around -- if you've marshalled it once, you're not going to marshall it again until you construct a new node. You may want to talk to @warpfork about this one.

Reply:
upon further further reflection it's not clear whether it's possible to remove this code from the go-ipfs PR :(, at least till we juse move everything over to the new stuff.

Reply:
Sounds like this is mostly to satisfy legacy interfaces, aka control the scope for this multi-repo merge party we've got coming up, and so it's here to stay, _for now_.

We can refresh ourselves on what the best path forward will be when we get another large block of work time to start chopping out legacy stuff.  (Sounds like we might end up in a position where we later try to remove several swaths at once, and it might end up easier that way.  But who knows.)  For now, I'd say "let's document the intent and move on", but even that seems pretty strongly hinted already by the func signature returning something called literally `legacy`!

So: I think we can call this resolved for now?  I'll mark it as such.